// azure-pipelines-language-service/lib/src/services/yamlValidation.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Red Hat, Inc. All rights reserved.
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var vscode_languageserver_types_1 = require("vscode-languageserver-types");
var jsonParser_1 = require("../parser/jsonParser");
var nls = require("vscode-nls");
var localize = nls.loadMessageBundle();
var YAMLValidation = /** @class */ (function () {
    function YAMLValidation(jsonSchemaService, promiseConstructor) {
        this.jsonSchemaService = jsonSchemaService;
        this.promise = promiseConstructor;
        this.validationEnabled = true;
    }
    YAMLValidation.prototype.configure = function (shouldValidate) {
        if (shouldValidate) {
            this.validationEnabled = shouldValidate.validate;
        }
    };
    YAMLValidation.prototype.doValidation = function (textDocument, yamlDocument) {
        if (!this.validationEnabled) {
            return this.promise.resolve([]);
        }
        if (yamlDocument.documents.length === 0) {
            //this is strange...
            return this.promise.resolve([]);
        }
        if (yamlDocument.documents.length > 1) {
            return this.promise.resolve([{
                    severity: vscode_languageserver_types_1.DiagnosticSeverity.Error,
                    range: {
                        start: {
                            line: 0,
                            character: 0
                        },
                        end: textDocument.positionAt(textDocument.getText().length)
                    },
                    message: localize('multiDocumentError', 'Only single-document files are supported')
                }]);
        }
        var translateSeverity = function (problemSeverity) {
            if (problemSeverity === jsonParser_1.ProblemSeverity.Error) {
                return vscode_languageserver_types_1.DiagnosticSeverity.Error;
            }
            if (problemSeverity == jsonParser_1.ProblemSeverity.Warning) {
                return vscode_languageserver_types_1.DiagnosticSeverity.Warning;
            }
            return vscode_languageserver_types_1.DiagnosticSeverity.Hint;
        };
        return this.jsonSchemaService.getSchemaForResource(textDocument.uri).then(function (schema) {
            var diagnostics = [];
            var jsonDocument = yamlDocument.documents[0];
            jsonDocument.errors.forEach(function (err) {
                diagnostics.push({
                    severity: vscode_languageserver_types_1.DiagnosticSeverity.Error,
                    range: {
                        start: textDocument.positionAt(err.start),
                        end: textDocument.positionAt(err.end)
                    },
                    message: err.getMessage()
                });
            });
            jsonDocument.warnings.forEach(function (warn) {
                diagnostics.push({
                    severity: vscode_languageserver_types_1.DiagnosticSeverity.Warning,
                    range: {
                        start: textDocument.positionAt(warn.start),
                        end: textDocument.positionAt(warn.end)
                    },
                    message: warn.getMessage()
                });
            });
            if (schema) {
                var added = {};
                var problems = jsonDocument.getValidationProblems(schema.schema);
                problems.forEach(function (problem, index) {
                    var message = problem.getMessage();
                    var signature = '' + problem.location.start + ' ' + problem.location.end + ' ' + message;
                    if (!added[signature]) {
                        added[signature] = true;
                        diagnostics.push({
                            severity: translateSeverity(problem.severity),
                            range: {
                                start: textDocument.positionAt(problem.location.start),
                                end: textDocument.positionAt(problem.location.end)
                            },
                            message: message
                        });
                    }
                });
                if (schema.errors.length > 0) {
                    for (var _i = 0, _a = schema.errors; _i < _a.length; _i++) {
                        var curDiagnostic = _a[_i];
                        diagnostics.push({
                            severity: vscode_languageserver_types_1.DiagnosticSeverity.Error,
                            range: {
                                start: {
                                    line: 0,
                                    character: 0
                                },
                                end: {
                                    line: 0,
                                    character: 1
                                }
                            },
                            message: curDiagnostic
                        });
                    }
                }
            }
            return diagnostics;
        });
    };
    return YAMLValidation;
}());
exports.YAMLValidation = YAMLValidation;
